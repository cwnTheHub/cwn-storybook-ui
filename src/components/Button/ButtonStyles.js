import styled from "styled-components";

export const ButtonContainer = styled.div`
  position: relative;
  width: 100%;
`;

export const ButtonField = styled.button`
  width: 100%;
  transition: background 0.2s ease 0s;
  background-color: ${({ btnBackgroundColor, outlined }) =>
    !outlined ? btnBackgroundColor : "rgba(216,216,216,0.2)"};
  color: ${({ btnTxtColor, outlined }) =>
    outlined ? "#000" : btnTxtColor}!important;
  margin: 0px;
  outline: 0px;
  text-overflow: ellipsis;
  border-color: ${({ btnBackgroundColor, outlined }) =>
    !outlined ? btnBackgroundColor : "transparent"};
  border-width: 1px;
  border-style: solid;
  border-radius: 4px !important;
  overflow-wrap: break-word;
  font-size: 1rem;
  letter-spacing: -0.8px;
  line-height: 1.5;
  font-weight: light;
  cursor: pointer;
  min-height: 1rem;
  min-width: 150px;
  padding: 0.3rem 1rem;
  box-sizing: border-box;

  &:hover {
    background-color: ${({ btnBackgroundColor, btnTxtColor }) =>
      btnBackgroundColor ? btnTxtColor : "transparent"};
    border-color: ${({ btnBackgroundColor, btnTxtColor }) =>
      btnBackgroundColor ? btnBackgroundColor : btnTxtColor};
    color: ${({ btnBackgroundColor, outlined, btnTxtColor }) =>
      outlined ? btnTxtColor : btnBackgroundColor} !important;
    border-width: ${({ outlined }) => (!outlined ? "2px" : "1px")};
    border-style: solid;
  }
`;
