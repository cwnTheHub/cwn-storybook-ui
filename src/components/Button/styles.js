import styled from "styled-components";

export const ButtonContainer = styled.div`
  position: relative;
  width: 100%;
`;

export const ButtonField = styled.button`
  width: ${({ ctaWidth }) => (ctaWidth ? ctaWidth : "180px")};
  height: ${({ ctaHeight }) => (ctaHeight ? ctaHeight : "3.25rem")};
  transition: background 0.2s ease 0s;

  background-color: ${({ ctaBgColor, variant }) =>
    variant == "outline"
      ? "rgba(216,216,216,0.2)"
      : variant
      ? ctaBgColor
      : "rgba(216,216,216,0.2)"};
  color: ${({ ctaTxtColor, variant, ctaBgColor }) =>
    variant == "outline"
      ? ctaBgColor
      : variant
      ? ctaTxtColor
      : "#000"}!important;
  margin: 0px;
  outline: 0px;
  text-overflow: ellipsis;
  border-color: ${({ ctaBgColor, variant }) =>
    variant ? ctaBgColor : "transparent"};
  border-width: ${({ variant }) => (variant ? "1px" : 0)};
  border-style: ${({ variant }) => variant && "solid"};
  border-radius: 4px !important;
  overflow-wrap: break-word;
  font-size: 1rem;
  letter-spacing: -0.8px;
  line-height: 1.5;
  font-weight: light;
  cursor: pointer;
  min-height: 1rem;
  min-width: 150px;
  padding: 0.3rem 1rem;
  box-sizing: border-box;

  &:hover {
    background-color: ${({ variant, ctaTxtColor }) =>
      variant === "outline"
        ? "rgba(216,216,216,0.2)"
        : variant
        ? ctaTxtColor
        : "#000"};
    border-color: ${({ ctaBgColor, ctaTxtColor }) =>
      ctaBgColor ? ctaBgColor : ctaTxtColor};
    color: ${({ ctaBgColor, variant }) =>
      variant ? ctaBgColor : "rgba(216,216,216,1)"} !important;
    border-width: ${({ variant }) => (variant ? "2px" : "1px")};
    border-style: solid;
  }
`;
