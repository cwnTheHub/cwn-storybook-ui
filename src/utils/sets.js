const difference = (start, compare) => {
  const setDifference = new Set(start);

  compare.forEach((element) => setDifference.delete(element));

  return setDifference;
};

export const isEqual = (setA, setB) => {
  const differenceAtoB = difference(setA, setB);
  const differenceBtoA = difference(setB, setA);

  return differenceAtoB.size === 0 && differenceBtoA.size === 0;
};
