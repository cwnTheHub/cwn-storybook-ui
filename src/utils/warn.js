
export const deprecate = (componentName, message) => {
  if (process.env.NODE_ENV === "production") {
    return;
  }

  console.warn(`[NDS] [Deprecate] ${componentName}: ${message}`); // eslint-disable-line no-console
};

export const warn = (componentName, message) => {
  if (process.env.NODE_ENV === "production") {
    return;
  }

  console.warn(`[NDS] ${componentName}: ${message}`); // eslint-disable-line no-console
};
