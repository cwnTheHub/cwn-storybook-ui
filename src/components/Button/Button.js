import React from "react";
import PropTypes from "prop-types";
import { ButtonContainer, ButtonField } from "./ButtonStyles";

const Button = ({
  outlined,
  btnBackgroundColor,
  btnTxtColor,
  onClick,
  disabled,
  buttonTxt
}) => {
  return (
    <ButtonContainer>
      <ButtonField
        data-testid="Click Me"
        onClick={onClick}
        disabled={disabled}
        btnBackgroundColor={btnBackgroundColor}
        btnTxtColor={btnTxtColor}
        outlined={outlined}
      >
        {buttonTxt}
      </ButtonField>
    </ButtonContainer>
  );
};

Button.propTypes = {
  icon: PropTypes.element,
  backgroundColor: PropTypes.string,
  btnTxtColor: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  buttonTxt: PropTypes.string,
  height: PropTypes.string,
  width: PropTypes.string,
};

export default Button;

