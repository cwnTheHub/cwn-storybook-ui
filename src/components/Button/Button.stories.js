import React from "react";
import Button from "./Button";

export default {
  title: "Button",
  component: Button,
};

const Template = (args) => {
  const {
    btnBackgroundColor,
    btnTxtColor,
    onClick,
    disabled,
    height,
    width,
    buttonTxt,
  } = args;
  return (
    <Button
      btnBackgroundColor={btnBackgroundColor}
      btnTxtColor={btnTxtColor}
      onClick={onClick}
      disabled={disabled}
      height={height}
      width={width}
      buttonTxt={buttonTxt}
    />
  );
};

export const Primary = Template.bind({});
Primary.args = {
  btnBackgroundColor: "green",
  btnTxtColor: "white",
  onClick: () => console.log("click"),
  disabled: false,
  height: "50px",
  width: "250px",
  buttonTxt: "Button",
};
