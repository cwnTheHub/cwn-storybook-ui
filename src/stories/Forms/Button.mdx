import {
  Canvas,
  Meta,
  Title,
  Subtitle,
  Description,
  Primary,
  Controls,
  Stories,
  Story,
} from "@storybook/blocks";

import * as ButtonStories from "./Button.stories";

<Meta of={ButtonStories} />
<Title />

`import { Button } from '@nds_core/nds'`

### PROPS & METHODS

<Controls />

## Minimal usage

Provide a function as the onClick prop to perform an action when clicked. **Avoid using a button if navigation is the primary action, as a Link is more appropriate.**

## Usage criteria

- Use buttons to move through a transaction
- Aim to use only one primary button variant per page
- Keep the text short and precise. It should fit on a single line (our recommendation is 2-24 characters for French and English)
- Make use of the A11yContent core component to provide more written context for assistive technology users for when they navigate a page using only button landmarks
- Make sure the button text describes an action
- Buttons should not be disabled. Use inline (on blur) error messaging to provide feedback when the form is invalid

Currently, by default, Buttons will be displayed in the Primary variant.

<Primary />

Specify the variant to create a button for secondary actions.

<Canvas of={ButtonStories.Secondary} />

### NEW Button variants and ranks

New Buttons and their associated colours are meant for different purposes. Choose different variants according to their context and rank.

### Ranks

By default, new Button variants will be displayed in the Common rank. Use Common rank for medium to low emphasis.

Use Main rank for the most emphasis and main action on a page or in a form.

### StandardCommon + StandardMain

Used mainly for marketing pages for commerce actions where the direct/immediate action is to purchase. It can also be used for login or registration as these will lead a user to pay their bill or add services.

<Canvas of={ButtonStories.StandardMain} />
<Canvas of={ButtonStories.StandardCommon} />

### BrandCommon + BrandMain

Used mainly for editorial, non-commerce-related actions where the direct/immediate action goes to a learn/info/editorial experience.

<Canvas of={ButtonStories.BrandCommon} />
<Canvas of={ButtonStories.BrandMain} />

### DangerCommon

Reserved for applications, native apps, and internal tools. Cancel of private/legal/monetary effect. If an action can be undone, do not use this.

<Canvas of={ButtonStories.Danger} />

### Sizes

All buttons are inline, with a minimum width of 180px for screens larger than 768px. They will occupy 100% width of their parent's width at the small viewport (≤ 767px) and below. Resize your browser window to see this behaviour.

### Placing buttons on dark backgrounds

Use the inverted button on top of a dark background.

Use this variant with caution. There will be accessibility issues if the colour contrast of the image and the button text is too low in the hover state.

<Canvas of={ButtonStories.Inverted} />

### Accessibility features

The Button component includes:

- High-contrast interactive states (default, active, focus) for all new button variants and ranks
- Makes use of the button element to provide contextual feedback to assistive technology
- Is compatible with the A11yContent core component to provide more written context to assistive technology without occupying too much visual space

<Canvas
  of={ButtonStories.StandardCommon}
/>

## Disabled buttons

Buttons should not be disabled on NDS experiences. The following is a list of reasons why:

- Disabled buttons fool users into thinking they can proceed with an action
- Disabled buttons are intentionally designed to be “hard to see”; however this is visually inaccessible
- Disabled buttons do not inherently give any feedback or reason why they are disabled
- They give design teams a reason to rush through error handling as opposed to providing the correct feedback to guide users to their next steps
- Disabled buttons may be cut off for low vision users who zoom-in on their devices causing confusion as to why they cannot proceed and important information to be missed
- Disabled buttons are usually not navigable with assistive technologies like screen readers and switches. This usually means that the disabled buttons will be skipped over

### What to do instead of using a disabled button:

If you show the button in its disabled state with no explanation, there can be confusion as to why it is disabled. To avoid user confusion, implement full error handling and feedback to your interactive experiences.

To better express to a user that they cannot proceed with form submission, show the submit Button, but perform error validation when the user clicks it, and display all errors as a group within an InputFeedback component adjacent to the submit button.

This allows for clear actions laid out for the user to take and should include skip links for easy navigation to parts of the form that need action

## Additional variants

<Stories />
