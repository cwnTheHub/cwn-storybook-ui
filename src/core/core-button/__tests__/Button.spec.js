import React from "react";
import { render, shallow } from "enzyme";

import Button from "../Button";
import A11yContent from "../../core-a11y-content/A11yContent";
import { warn } from "../../../utils/warn";

jest.mock("../../../utils/warn");

describe("Button", () => {
  const doShallow = (overrides = {}) => {
    const button = shallow(<Button {...overrides}>Submit</Button>);

    return button;
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders", () => {
    const button = render(<Button>Submit</Button>);

    expect(button).toMatchSnapshot();
  });

  it("has one of the HTML button types", () => {
    let button = doShallow();
    expect(button.props().type).toBe("button");

    button = doShallow({ type: "reset" });
    expect(button.props().type).toBe("reset");
  });

  it("can be presented as one of the allowed variants", () => {
    const button = render(<Button variant="secondary">Submit</Button>);

    expect(button).toMatchSnapshot();
  });

  it("can be presented as one of the new allowed variants", () => {
    const button = render(<Button variant="standard">Submit</Button>);

    expect(button).toMatchSnapshot();
  });

  it("can not be disabled", () => {
    const button = doShallow({ disabled: true });

    expect(button.props()["disabled"]).toBe(undefined);
    expect(warn).toHaveBeenCalled();
  });

  it("passes additional attributes to button element", () => {
    const button = doShallow({ id: "the-button", tabIndex: 1 });

    expect(button.props().id).toBe("the-button");
    expect(button.props()["tabIndex"]).toBe(1);
  });

  describe("A11yContent", () => {
    it("connects to Button", () => {
      const button = shallow(
        <Button>
          Go home
          <A11yContent>testing</A11yContent>
        </Button>
      );

      expect(button.contains(<A11yContent>testing</A11yContent>)).toBe(true);
    });
  });

  it("does not allow custom CSS", () => {
    const button = doShallow({
      className: "my-custom-class",
      style: { color: "hotpink" },
    });

    expect(button.props()["className"]).not.toEqual("my-custom-class");
    expect(button.props()["style"]).toBe(undefined);
  });
});
