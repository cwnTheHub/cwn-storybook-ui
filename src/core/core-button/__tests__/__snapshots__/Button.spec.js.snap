// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button can be presented as one of the allowed variants 1`] = `
<button
  class="sc-beqWaB etEeBB"
  type="button"
>
  <span
    class="sc-gueYoa htAACu"
  >
    Submit
  </span>
</button>
`;

exports[`Button can be presented as one of the new allowed variants 1`] = `
<button
  class="sc-beqWaB dWrusy"
  type="button"
>
  <span
    class="sc-gueYoa bJKfsr"
  >
    Submit
  </span>
</button>
`;

exports[`Button renders 1`] = `
<button
  class="sc-beqWaB iJOUxQ"
  type="button"
>
  <span
    class="sc-gueYoa htAACu"
  >
    Submit
  </span>
</button>
`;
